### ISSUE TITLE
Quoted arguments with spaces propagated incorrectly to spawned process

### ISSUE THREAD

**Issue by atifaziz** [atifaziz/lshim#1]
The process spawned by `elvee` sees quoted arguments with spaces as separate arguments such that arguments `foo`, `"bar baz"` and `qux` will be passed as `foo`, `bar`, `baz`, and `qux`.

## Steps to Reproduce

Download [version 1.0](https://github.com/atifaziz/elvee/releases/tag/v1.0.0) [for Windows](https://github.com/atifaziz/elvee/releases/download/v1.0.0/elvee-1.0.0-win-x64.zip) to `%temp%`.

Assuming `%temp%` is the current working directory (i.e. `cd %temp%`), run `mkdir v1` to create a sub-directory to hold some version 1 of a batch script.

Next, create a batch `v1\foo.cmd` in that sub-directory (i.e. full path being `%temp%\v1\foo.cmd`) and set its content to:

```bat
@echo off
setlocal
set X=1
:loop
if [%1]==[] goto :EOF
echo [%X%] = %1
shift
set /a X=X+1
goto :loop
```

The above batch script simply dumps its arguments, one per line, while numbering them.

First run:

    elvee .\?\foo.cmd bar baz qux quux

where `foo` is called with 4 arguments.

Next run:

    elvee .\?\foo.cmd bar "baz qux" quux

where `foo` is called with 3 arguments (`baz qux` being now a single quoted argument).

## Expected Result

Expected first run to output:

```
[1] = bar
[2] = baz
[3] = qux
[4] = quux
```

and second run to output:

```
[1] = bar
[2] = baz qux
[3] = quux
```

## Actual Result

Instead outputs for both runs are identical. The second run's output is therefore incorrect. It received 4 arguments:

```
[1] = bar
[2] = baz
[3] = qux
[4] = quux
```

when it should have received 3. Quoting an argument makes no difference and embedded spaces continue to delimit arguments.

## More Information

This was observed on Windows only so it's not clear if the issue exists for other platforms too.

**Comment by atifaziz** [comment]
The issue seems to stem from the use of `_spawnv`:

https://github.com/atifaziz/elvee/blob/0a9f23c1282b9f2bf474bdf8d72c310874145213/elvee.c#L306

which according to [the documentation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/spawn-wspawn-functions?view=vs-2019#arguments-for-the-spawned-process) causes unexpected behaviour if an argument contains embedded spaces:

> **Note**
>
> Spaces embedded in strings may cause unexpected behavior; for example, passing `_spawn` the string `"hi there"` will result in the new process getting two arguments, `"hi"` and `"there"`. If the intent was to have the new process open a file named `"hi there"`, the process would fail. You can avoid this by quoting the string: `"\"hi there\""`.

**Comment by atifaziz** [comment]
I just posted a [bounty] 💰 if anyone would like to help fix this bug. Hopefully it'll pay for a nice :pizza: + :beer: (or in some countries you could feed a whole family) as gratitude.

---
[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=86627967)][bounty]


[bounty]: https://www.bountysource.com/issues/86627967-quoted-arguments-with-spaces-propagated-incorrectly-to-spawned-process

**Comment by atifaziz** [comment]
See “[Everyone quotes command line arguments the wrong way](https://docs.microsoft.com/en-us/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way)” for good information on how to solve this correctly.

**Comment by JaniM** [comment]
A quick looksy revealed that the code is ok on Linux as `execv` replaces the process - ie, it can use the same memory and we have no need to serialize anything. Why the similarly named `_execv` on Windows is an entirely different beast is beyond me.

Opened a PR. #2

**Comment by atifaziz** [comment]
> Why the similarly named `_execv` on Windows is an entirely different beast is beyond me.

That's because there is no fork-exec model on Windows. It has a [distinct API for creating another process](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa) with an explicitly passed environment, arguments and options for controlling the launch. I suspect that `_execv` is just provided for some help with porting POSIX applications. The compatibility note says that it's not fully conforming:

> The UCRT also implements a large subset of the POSIX.1 (ISO/IEC 9945-1:1996, the POSIX System Application Program Interface) C library. However, it's not fully conformant to any specific POSIX standard. The UCRT also implements several Microsoft-specific functions and macros that aren't part of a standard.

**Comment by JaniM** [comment]
Yep, I understand the architectural differences, it's just amusing to see these "compatibility" functions that are just very likely to lead to headaches.

**Comment by atifaziz** [comment]
> it's just amusing to see these "compatibility" functions that are just very likely to lead to headaches.

I hear ya.

**Comment by atifaziz** [comment]
Closed via 211bd1edf39dd25c088bb825276e6a2b3f3b48a2, which concludes of PR #2.

**Comment by atifaziz** [comment]
The fix is now released in [version 1.1](https://github.com/atifaziz/elvee/releases/tag/v1.1.0).
